### PR TITLE
New Browsing Menu

### PR DESCRIPTION
**Pull request by bwaresiak**
Task/Issue URL: https://app.asana.com/0/856498667320406/1199661417922878
Tech Design URL: https://app.asana.com/0/856498667320406/1199971443872605
CC:

**Description**:
- Introduce new Menu.
- Change Home Screen buttons: swap Bookmarks and Tabs.
- When browsing:
  - Expose Share button on the Top Bar.
  - Move Menu button to Bottom Bar.
 - Introduce new Toast view with action button.

**Steps to test this PR**:

1. Full Menu UX & features test.
  - Sizing, layout, animations.
  - Test each functionality.
  - Home screen & Share buttons.
  - iPhone & iPad.
  - Older iOS versions for possible layout issues.
2. Test new Toast.
3. Add-favorites flow from widget.

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPad

**OS Testing**:

* [ ] iOS 11
* [ ] iOS 12
* [ ] iOS 13
* [ ] iOS 14

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

### CODE REVIEW

File: DuckDuckGo/BrowsingMenu/BrowsingMenuViewController.swift [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+        case .regular(_, _, _, let action):
+            dismiss?()
+            action()
+        case .separator:
+            break
+        }
+    }
+}
+
+// swiftlint:disable line_length
```
Comment (brindy) [inline DuckDuckGo/BrowsingMenu/BrowsingMenuViewController.swift:294]: Seems unnecessary?

File: DuckDuckGo/BrowsingMenu/BrowsingMenuViewController.swift [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+        }
+    }
+    
+    func flashScrollIndicatorsIfNeeded() {
+        if tableView.bounds.height < tableView.contentSize.height {
+            tableView.flashScrollIndicators()
+        }
+    }
+    
+    func setHeaderEntires(_ entries: [BrowsingMenuEntry]) {
```
Comment (brindy) [inline DuckDuckGo/BrowsingMenu/BrowsingMenuViewController.swift]: typo: `Entires` -> `Entries`

File: DuckDuckGo/BrowsingMenu/BrowsingMenuViewController.swift [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+            view.configure(with: image, label: name, accessibilityLabel: accessibilityLabel) { [weak self] in
+                self?.dismiss?()
+                action()
+            }
+        }
+        
+        headerEntries = entries
+    }
+    
+    func setMenuEntires(_ entries: [BrowsingMenuEntry]) {
```
Comment (brindy) [inline DuckDuckGo/BrowsingMenu/BrowsingMenuViewController.swift]: typo again, at least it's consistent :)

File: DuckDuckGo/BrowsingMenu/BrowsingMenuViewController.swift [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import UIKit
+
+protocol BrowsingMenu {
+    
+    func setMenuEntires(_ entries: [BrowsingMenuEntry])
```
Comment (brindy) [inline DuckDuckGo/BrowsingMenu/BrowsingMenuViewController.swift]: typo: `Entires` -> `Entries`

File: DuckDuckGo/TabViewControllerBrowsingMenuExtension.swift [hunk 5]
```diff
@@ -22,108 +22,204 @@ import Core
 
 extension TabViewController {
     
-    func buildBrowsingMenu() -> UIAlertController {
+    func buildBrowsingMenuHeaderContent() -> [BrowsingMenuEntry] {
         
-        let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
-        alert.overrideUserInterfaceStyle()
-        alert.addAction(title: UserText.actionNewTab) { [weak self] in
+        var entires = [BrowsingMenuEntry]()
```
Comment (brindy) [inline DuckDuckGo/TabViewControllerBrowsingMenuExtension.swift]: typo: `entires` -> `entries`

**Review by brindy — APPROVED** [review]
Generally looking good.  Tests well. Works great on iOS 11, iPad, etc.  All flows seem good.

A few comments inline, nothing major. 

Additionally, is it worth removing the word "Toast" from function/var names, especially now we have the "ActionMessage" which is a much nicer name?

Finally, I made some comments in the product review but assuming none of those end up changing the code, feel free to merge once the above comments are addressed.

File: DuckDuckGo/BrowsingMenu/BrowsingMenuViewController.swift (hunk 1 above)
Comment (bwaresiak) [reply DuckDuckGo/BrowsingMenu/BrowsingMenuViewController.swift:294]: No, it's needed for guards :)

**Comment by bwaresiak** [comment]
@brindy I made one fix to a scenario that Craig found:
1. Enable clear data, with remove tabs option after 5 min.
2. Leave the app, wait 5min.
3. Open any URL (default browser flow) - menu button was not visible.

**Review by samsymons — CHANGES_REQUESTED** [review]
This is such good work here, nice one @bwaresiak. I'm spending some time stress testing this and collecting feedback below:

---

* If you display the menu and then use Command+T to create a new tab while it is still visible, it breaks the layout
* `BrowsingMenuButton` behaves strangely as an accessibility element. Likely it should be treated as `.button` accessibility element and hide its children from VoiceOver

new-tab-layout

**Comment by bwaresiak** [comment]
@samsymons Thanks!

> If you display the menu and then use Command+T to create a new tab while it is still visible, it breaks the layout

Fixed, along with `cmd+w` and `cmd+shift+[` scenarios.

> BrowsingMenuButton behaves strangely as an accessibility element. Likely it should be treated as .button accessibility element and hide its children from VoiceOver

But it is .button for me... but I've found that accessibility labels for Bookmark vs Menu states were broken, so I've fixed that. Can you retest?

**Comment by samsymons** [comment]
@bwaresiak This looks good now. I was having a hard time navigating to the 4 buttons at the top with VoiceOver and I wasn't seeing them as a button with the Accessibility Inspector, but it's working now. 🎉

**Review by samsymons — APPROVED** [review]

**Comment by bwaresiak** [comment]
Thank you @samsymons  and @brindy

**Pull request merged**